### PR TITLE
Support for dynamic allocation for spark 1.6.1

### PR DESCRIPTION
**Pull request by solnic**

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/CoarseCookSchedulerBackend.scala [hunk 1]
```diff
@@ -115,20 +118,37 @@ private[spark] class CoarseCookSchedulerBackend(
...
+          logWarning(s"Job ${job.getUUID} has died. Failure ($totalFailures/$maxFailures)")
+
+          if (totalFailures >= maxFailures) {
+            // TODO should we abort the outstanding tasks now?
+            logError(s"We have exceeded our maximum failures ($maxFailures)" +
+              "and will not relaunch any more tasks")
+          }
+        }
+
+        runningJobUUIDs = runningJobUUIDs - uuid
```
Comment (wyegelwel) [inline core/src/main/scala/org/apache/spark/scheduler/CoarseCookSchedulerBackend.scala]: Should this be in the `isCompleted` branch?

Comment (solnic) [inline core/src/main/scala/org/apache/spark/scheduler/CoarseCookSchedulerBackend.scala]: You're right, I changed it (along with a couple other fixes). More info in my update on basecamp :)

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 2]
```diff
@@ -15,23 +15,26 @@
...
+package org.apache.spark.scheduler.cluster.cook
 
 import java.io.{ BufferedWriter, FileWriter }
 import java.net.URI
 import java.nio.file.{ Files, Paths }
-import java.util.UUID
+import java.util.{ UUID, List => JList }
 
 import scala.collection.JavaConverters._
+import scala.collection.mutable.{ HashMap }
```
Comment (leifwalsh) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: Doesn't need braces

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 3]
```diff
@@ -115,20 +118,34 @@ private[spark] class CoarseCookSchedulerBackend(
...
-        runningJobUUIDs = runningJobUUIDs - job.getUUID
-        if (totalFailures >= maxFailures) {
-          // TODO should we abort the outstanding tasks now?
-          logError(s"We have exceeded our maximum failures ($maxFailures)" +
-                    "and will not relaunch any more tasks")
-        } else requestRemainingCores()
+        abortedJobIds = abortedJobIds - uuid
+        runningJobUUIDs = runningJobUUIDs - uuid
+
+        if (!job.isSuccess && !isAborted) {
```
Comment (leifwalsh) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: Might be nice to log something when we notice that an aborted job was successfully aborted, either at INFO or DEBUG level

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 4]
```diff
@@ -182,33 +206,57 @@ private[spark] class CoarseCookSchedulerBackend(
 
     val cmds = remoteConfFetch ++ environment ++ Seq(commandString, cleanup)
 
-    new Job.Builder()
+    val job = new Job.Builder()
```
Comment (leifwalsh) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: Why add this local variable just to return it?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 5]
```diff
@@ -182,33 +206,57 @@ private[spark] class CoarseCookSchedulerBackend(
...
+      .setMemory(mesosBackend.calculateTotalMemory(sc))
       .setCpus(numCores)
       .setPriority(priority)
       .build()
+
+    job
   }
 
   def createRemainingJobs(): List[Job] = {
     def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
```
Comment (leifwalsh) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: Should have tailrec annotation

**Comment by leifwalsh** [comment]
Otherwise, LGTM.  After making those fixes and approval from Wenbo, please squash before merging.

**Comment by solnic** [comment]
@leifwalsh hey Leif :) thanks for the review, I've addressed the issues you mentioned

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 6]
```diff
@@ -15,23 +15,28 @@
...
 import org.apache.mesos.Protos._
 
 import org.apache.spark.{Logging, SparkContext}
+
+import org.apache.spark.scheduler.TaskSchedulerImpl
 import org.apache.spark.scheduler.cluster.CoarseGrainedSchedulerBackend
 import org.apache.spark.scheduler.cluster.mesos.CoarseMesosSchedulerBackend
 
-import com.twosigma.cook.jobclient.{Job, FetchableURI, JobClient, JobListener => CJobListener}
+import com.twosigma.cook.jobclient.{Job, FetchableURI, JobClient, JobClientException, JobListener => CJobListener}
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: missing space `{` and `}` or make the previous one consistent

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 7]
```diff
@@ -187,28 +215,50 @@ private[spark] class CoarseCookSchedulerBackend(
...
+     @tailrec def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
       if (coresRemaining <= 0) jobs
       else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
       else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
-    loop(maxCores - totalCoresRequested, Nil).reverse
+
+    loop(currentCoresLimit, List()).reverse
+  }
+
+  def currentCoresLimit(): Int = {
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]:
```
private def currentCoresLimit(): Int = jobLimitOption.map(_ * maxCoresPerJob).getOrElse(maxCores) - totalCoresRequested
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 8]
```diff
@@ -187,28 +215,50 @@ private[spark] class CoarseCookSchedulerBackend(
...
   }
 
   def requestRemainingCores() : Unit = {
     val jobs = createRemainingJobs()
+
     totalCoresRequested += jobs.map(_.getCpus.toInt).sum
     runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
-    jobClient.submit(jobs.asJava, jobListener)
+
+    if (!jobs.isEmpty) jobClient.submit(jobs.asJava, jobListener)
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: miss the `{}`

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 9]
```diff
@@ -219,7 +269,21 @@ private[spark] class CoarseCookSchedulerBackend(
 
   override def stop() {
     super.stop()
-    jobClient.abort(runningJobUUIDs.asJava)
+    abortJobs(runningJobUUIDs.toList)
+  }
+
+  private def abortJobs(uuids: List[UUID]) : Boolean = {
+    try {
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: Should use scala.try instead

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 10]
```diff
@@ -187,28 +215,50 @@ private[spark] class CoarseCookSchedulerBackend(
...
-      .setMemory(sparkMesosScheduler.calculateTotalMemory(sc))
+      .setMemory(mesosBackend.calculateTotalMemory(sc))
       .setCpus(numCores)
       .setPriority(priority)
       .build()
   }
 
   def createRemainingJobs(): List[Job] = {
-    def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
+     @tailrec def loop(coresRemaining: Int, jobs: List[Job]): List[Job] =
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: Should be different line

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 11]
```diff
@@ -187,28 +215,50 @@ private[spark] class CoarseCookSchedulerBackend(
...
       if (coresRemaining <= 0) jobs
       else if (coresRemaining <= maxCoresPerJob) createJob(coresRemaining) :: jobs
       else loop(coresRemaining - maxCoresPerJob, createJob(maxCoresPerJob) :: jobs)
-    loop(maxCores - totalCoresRequested, Nil).reverse
+
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: remove extra line

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 12]
```diff
@@ -187,28 +215,50 @@ private[spark] class CoarseCookSchedulerBackend(
...
     val jobs = createRemainingJobs()
+
     totalCoresRequested += jobs.map(_.getCpus.toInt).sum
     runningJobUUIDs = runningJobUUIDs ++ jobs.map(_.getUUID)
-    jobClient.submit(jobs.asJava, jobListener)
+
+    if (!jobs.isEmpty) jobClient.submit(jobs.asJava, jobListener)
+  }
+
+  override def doKillExecutors(executorIds: Seq[String]): Boolean = {
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: Should be just one line

```
override def doKillExecutors(executorIds: Seq[String]): Boolean = abortJobs(executorIds.map(x => UUID.fromString(x)).toList)
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 13]
```diff
@@ -187,28 +215,50 @@ private[spark] class CoarseCookSchedulerBackend(
...
+    val uuids = executorIds.map(x => UUID.fromString(x)).toList
+    abortJobs(uuids)
   }
 
-  // TODO we should consider making this async, because if there are issues with cook it'll
+  override def doRequestTotalExecutors(requestedTotal: Int): Boolean = {
+    logInfo("Capping the total amount of executors to " + requestedTotal)
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: `s"Capping the total amount of executors to $requestedTotal"`

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 14]
```diff
@@ -115,20 +120,36 @@ private[spark] class CoarseCookSchedulerBackend(
...
-                    "and will not relaunch any more tasks")
-        } else requestRemainingCores()
+        abortedJobIds = abortedJobIds - uuid
+        runningJobUUIDs = runningJobUUIDs - uuid
+
+        if (isAborted) logInfo(s"Job $uuid has been successfuly aborted")
+
+        if (!job.isSuccess && !isAborted) {
+          totalFailures += 1
+          logWarning(s"Job ${uuid} has died. Failure ($totalFailures/$maxFailures)")
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: ${uuid} => $uuid. And there are a lot of other places which need to unify

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 15]
```diff
@@ -115,20 +120,36 @@ private[spark] class CoarseCookSchedulerBackend(
...
-        runningJobUUIDs = runningJobUUIDs - job.getUUID
-        if (totalFailures >= maxFailures) {
-          // TODO should we abort the outstanding tasks now?
-          logError(s"We have exceeded our maximum failures ($maxFailures)" +
-                    "and will not relaunch any more tasks")
-        } else requestRemainingCores()
+        abortedJobIds = abortedJobIds - uuid
+        runningJobUUIDs = runningJobUUIDs - uuid
+
+        if (isAborted) logInfo(s"Job $uuid has been successfuly aborted")
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: always add `{}` for if-else

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 16]
```diff
@@ -115,20 +120,36 @@ private[spark] class CoarseCookSchedulerBackend(
   }
 
   var runningJobUUIDs = Set[UUID]()
+  var abortedJobIds = Set[UUID]()
+
+  private var jobLimitOption: Option[Int] = None
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: Could you add a doc to explain what it is? Also, I would prefer remove `Option` from the var name.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 17]
```diff
@@ -115,20 +120,36 @@ private[spark] class CoarseCookSchedulerBackend(
...
+        if (isAborted) logInfo(s"Job $uuid has been successfuly aborted")
+
+        if (!job.isSuccess && !isAborted) {
+          totalFailures += 1
+          logWarning(s"Job ${uuid} has died. Failure ($totalFailures/$maxFailures)")
+
+          if (totalFailures >= maxFailures) {
+            // TODO should we abort the outstanding tasks now?
+            logError(s"We have exceeded our maximum failures ($maxFailures)" +
+                        "and will not relaunch any more tasks")
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: The indent doesn't look right to me.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 18]
```diff
@@ -58,6 +58,7 @@ import org.apache.spark.rpc.RpcEndpointRef
 import org.apache.spark.scheduler._
 import org.apache.spark.scheduler.cluster.{CoarseGrainedSchedulerBackend,
   SparkDeploySchedulerBackend, SimrSchedulerBackend}
+import org.apache.spark.scheduler.cluster.cook.CoarseCookSchedulerBackend
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:61]: Why we need this?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala [hunk 19]
```diff
@@ -187,28 +215,50 @@ private[spark] class CoarseCookSchedulerBackend(
...
+    val uuids = executorIds.map(x => UUID.fromString(x)).toList
+    abortJobs(uuids)
   }
 
-  // TODO we should consider making this async, because if there are issues with cook it'll
+  override def doRequestTotalExecutors(requestedTotal: Int): Boolean = {
+    logInfo("Capping the total amount of executors to " + requestedTotal)
+    jobLimitOption = Some(requestedTotal)
+    requestRemainingCores
+    true
```
Comment (WenboZhao) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: We always return true? When should we return false?

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 18 above)
Comment (solnic) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:61]: It's needed by `createTaskScheduler` in SparkContext

File: core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala (hunk 19 above)
Comment (solnic) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/cook/CoarseCookSchedulerBackend.scala]: This is the required behavior for dyn-alloc. It works the same in [mesos scheduler backend](https://github.com/mojotech/spark/blob/cook-scheduler/core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala#L402). If this method doesn't return `true` then ExecutorAllocationManager [will warn](https://github.com/mojotech/spark/blob/cook-scheduler/core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala#L358-L373) that it didn't manage to reach the cluster manager to request executors.

**Comment by solnic** [comment]
@WenboZhao I've addressed your comments, thanks for reviewing :)

**Comment by WenboZhao** [comment]
LGTM, please squash before merging.

**Pull request merged**